crates/ruma-events/src/call/member/member_event.rs

```diff
@@ -0,0 +1,217 @@
+    pub fn active_memberships(
+        &self,
+        origin_server_ts: Option<MilliSecondsSinceUnixEpoch>,
+    ) -> Vec<MembershipData<'_>> {
+        match self {
+            CallMemberEventContent::LegacyContent(content) => content
+                .memberships
+                .iter()
+                .filter(|m| !m.is_expired(origin_server_ts))
+                .map(MembershipData::Legacy)
+                .collect(),
+            CallMemberEventContent::SessionContent(content) => {
+                [content].map(MembershipData::Session).to_vec()
+            }
+            CallMemberEventContent::Empty { leave_reason: _ } => Vec::new(),
+        }
+    }
+
+    /// All the memberships for this event. Can only contain multiple elements in the case of legacy
+    /// `m.call.member` state events.
+    pub fn memberships(&self) -> Vec<MembershipData<'_>> {
+        match self {
+            CallMemberEventContent::LegacyContent(content) => {
+                content.memberships.iter().map(MembershipData::Legacy).collect()
+            }
+            CallMemberEventContent::SessionContent(content) => {
+                [content].map(MembershipData::Session).to_vec()
+            }
+            CallMemberEventContent::Empty { leave_reason: _ } => Vec::new(),
+        }
+    }
+
+    /// Set the `created_ts` of each [`MembershipData::Legacy`] in this event.
+    ///
+    /// Each call member event contains the `origin_server_ts` and `content.create_ts`.
+    /// `content.create_ts` is undefined for the initial event of a session (because the
+    /// `origin_server_ts` is not known on the client).
+    /// In the rust sdk we want to copy over the `origin_server_ts` of the event into the
+    /// (This allows to use `MinimalStateEvents` and still be able to determine if a
+    /// expired)
+    pub fn set_created_ts_if_none(&mut self, origin_server_ts: MilliSecondsSinceUnixEpoch) {
+        match self {
+            CallMemberEventContent::LegacyContent(content) => {
+                content.memberships.iter_mut().for_each(|m: &mut LegacyMembershipData| {
+                    m.created_ts.get_or_insert(origin_server_ts);
+                });
+            }
+            CallMemberEventContent::SessionContent(m) => {
+                m.created_ts.get_or_insert(origin_server_ts);
+            }
+            _ => (),
+        }
+    }
+}
+
+impl RedactContent for CallMemberEventContent {
+    type Redacted = RedactedCallMemberEventContent;
+
+    fn redact(self, _version: &ruma_common::RoomVersionId) -> Self::Redacted {
+        RedactedCallMemberEventContent {}
+    }
+}
+
+/// The PossiblyRedacted version of [`CallMemberEventContent`].
+///
+/// This wraps the [`CallMemberEventContent`] since the redacted state
+/// Empty {} case is a valid/"meaningful" state event content.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[allow(clippy::exhaustive_structs)]
+pub struct PossiblyRedactedCallMemberEventContent(pub CallMemberEventContent);
```

Comment:
Is this the way to go?
Making Possibly redact contain the `CallMemberEventContent` enum that can be `Empty {}`

How would ruma decide if its a redacted event or a PossiblyRedacted event that has an empty event body?

Reply:
I believe that you could just declare it as a type alias:

```rust
pub type PossiblyRedactedCallMemberEventContent = CallMemberEventContent;
```

`Redacted*` and `PossiblyRedacted*` are not used in the same context. `PossiblyRedacted*` is actually used in places where we don't know if the event was redacted or not.

Reply:
It's actually used in `prev_content`. Since we only have the content and not the whole event, we don't have the `unsigned` field to know if it was redacted.

Reply:
So for prev_event for instance. This all makes sense now

Reply:
Would it also make sense to do the same with the redacted content? Or is it better to be explixit in this case?
Redacted content could never be `CallMemberEventContent::Session(...)`

Reply:
No, it's better if `Redacted*` only represents the possible states of a redacted event.